akka-cluster/src/main/scala/akka/cluster/CrossDcClusterHeartbeat.scala

```diff
@@ -0,0 +1,321 @@
+    /** Since we need ordering of oldests guaranteed, we must only look at Up (or Leaving, Exiting...) nodes */
+    def atLeastInUpState(m: Member): Boolean =
+      m.status != MemberStatus.WeaklyUp && m.status != MemberStatus.Joining
+
+    val selfDcNeighbours: SortedSet[Member] = state.getOrElse(
+      selfDc,
+      throw new IllegalStateException(s"Self DataCenter was [$selfDc] however no such entries in ${Logging.simpleName(getClass)} present. " +
+        s"This is very likely a bug, please turn on debug logging and open an issue about it."))
+    val selfDcOldOnes = selfDcNeighbours.filter(atLeastInUpState).take(nrOfMonitoredNodesPerDc)
+
+    // if this node is part of the "n oldest nodes" it should indeed monitor other nodes:
+    val shouldMonitorActively = selfDcOldOnes.exists(_.uniqueAddress == selfAddress)
+    shouldMonitorActively
+  }
+
+  def addMember(m: Member): CrossDcHeartbeatingState = {
+    val dc = m.dataCenter
+    def addMember0(member: Member): CrossDcHeartbeatingState = {
+      val updatedMembers = state.getOrElse(dc, emptyMembersSortedSet) + m
+      val updatedState = this.copy(state = state.updated(dc, updatedMembers))
+
+      // guarding against the case of two members having the same upNumber, in which case the activeReceivers
+      // which are based on the ageOrdering could actually have changed by adding a node. In practice this 
+      // should happen rarely, since upNumbers are assigned sequentially, and we only ever compare nodes 
+      // in the same DC. If it happens though, we need to remove the previously monitored node from the failure 
+      // detector to prevent a resource leak
+      val stoppedMonitoringReceivers = updatedState.activeReceiversIn(dc) -- this.activeReceiversIn(dc)
+      stoppedMonitoringReceivers.foreach(m ⇒ failureDetector.remove(m.address)) // at most one element difference 
+
+      updatedState
+    }
+
+    // we need to remove the member first, to avoid having "duplicates"
+    // this is because the removal and uniqueness we need is only by uniqueAddress
+    // which is not used by the `ageOrdering`
+    removeMember(m)
+    addMember0(m)
+  }
+
+  def removeMember(m: Member): CrossDcHeartbeatingState = {
+    val dc = m.dataCenter
+    state.get(dc) match {
+      case Some(dcMembers) ⇒
+        val updatedMembers = dcMembers.filterNot(_.uniqueAddress == m.uniqueAddress)
+
+        failureDetector.remove(m.address)
+        copy(state = state.updated(dc, updatedMembers))
+      case None ⇒
+        this // no change needed, was certainly not present (not even its DC was) 
+    }
+  }
+
+  val activeReceivers: Set[UniqueAddress] =
+    Set() ++ dataCenters.flatMap(k ⇒ state(k).take(nrOfMonitoredNodesPerDc).map(_.uniqueAddress).toSet)
+
+  private def activeReceiversIn(dc: DataCenter): Set[UniqueAddress] =
+    Set() ++ state.getOrElse(dc, emptyMembersSortedSet).take(nrOfMonitoredNodesPerDc).map(_.uniqueAddress).toSet
```

Comment:
breakOut saves the day <3

`state.getOrElse(dc, emptyMembersSortedSet).take(nrOfMonitoredNodesPerDc).map(_.uniqueAddress)(breakOut)`